test/cpp/src/701-dram-scheduler.cc

```diff
@@ -0,0 +1,116 @@
+#include <catch.hpp>
+#include "mocks.hpp"
+#include "defaults.hpp"
+#include "dram_controller.h"    
+#include "champsim_constants.h"
+#include <algorithm>
+#include <cfenv>
+#include <cmath>
+
+bool dram_test(MEMORY_CONTROLLER* uut, std::vector<champsim::channel::request_type>* packet_stream, std::vector<uint64_t>* expected_order, std::vector<uint64_t>* arriv_time)
+{
+    uint64_t vector_index = 0;
+    uut->current_cycle = 0;
+
+    std::vector<DRAM_CHANNEL::queue_type::iterator> request_locations;
+    //load requests into controller
+    for(auto it = packet_stream->begin(); it != packet_stream->end(); it++)
+    {
+        auto jt = uut->channels[0].RQ.begin() + vector_index;
+        *jt = DRAM_CHANNEL::request_type{*it};
+        jt->value().forward_checked = false;
+        jt->value().event_cycle = (*arriv_time)[vector_index];
+        request_locations.push_back(jt);
+        vector_index++;
+    }
```

Comment:
It would be safer to use a transformation here rather than using pointer arithmetic directly.

```suggestion
    auto ins_begin = std::begin(uut->channels[0].RQ);
    //load requests into controller
    std::transform(std::cbegin(*packet_stream), std::cend(*packet_stream), std::cbegin(*arriv_time), ins_begin, [](auto pkt, uint64_t cycle) {
        pkt.forward_checked = false;
        pkt.event_cycle = cycle;
        return DRAM_CHANNEL::request_type{pkt};
    });

    std::vector<DRAM_CHANNEL::queue_type::iterator> request_locations(std::size(*packet_stream));
    std::iota(std::cbegin(request_locations), std::cend(request_locations), ins_begin);
```